theblues/charmstore.py

```diff
@@ -118,67 +127,90 @@ def entities(self, entity_ids):
+        return url
 
-    def charm_icon(self, charm_id):
-        url = self.charm_icon_url(charm_id)
+    def charm_icon(self, charm_id, channel=None):
+        url = self.charm_icon_url(charm_id, channel=channel)
         response = self._get(url)
         return response.content
 
-    def bundle_visualization(self, bundle_id):
-        url = self.bundle_visualization_url(bundle_id)
+    def bundle_visualization(self, bundle_id, channel=None):
+        url = self.bundle_visualization_url(bundle_id, channel=channel)
         response = self._get(url)
         return response.content
 
-    def bundle_visualization_url(self, bundle_id):
-        return '%s/%s/diagram.svg' % (self.url, _get_path(bundle_id))
+    def bundle_visualization_url(self, bundle_id, channel=None):
+        url = '{}/{}/diagram.svg'.format(self.url, _get_path(bundle_id))
+        if channel is not None:
+            url = '{}?channel={}'.format(url, channel)
+        return url
 
-    def entity_readme_url(self, entity_id):
+    def entity_readme_url(self, entity_id, channel=None):
         '''Generate the url path for the readme of the entity.'''
-        return '%s/%s/readme' % (self.url, _get_path(entity_id))
+        url = '{}/{}/readme'.format(self.url, _get_path(entity_id))
+        if channel is not None:
+            url = '{}?channel={}'.format(url, channel)
+        return url
 
-    def entity_readme_content(self, entity_id):
-        readme_url = self.entity_readme_url(entity_id)
+    def entity_readme_content(self, entity_id, channel=None):
+        readme_url = self.entity_readme_url(entity_id, channel=channel)
         response = self._get(readme_url)
         return response.text
 
-    def archive_url(self, entity_id):
+    def archive_url(self, entity_id, channel=None):
         '''Generate a URL for the archive.
 
         @param entity_id The ID of the entity to look up as a string
                or reference.
+        @param channel optional channel information.
         '''
-        return '%s/%s/archive' % (self.url, _get_path(entity_id))
+        url = '{}/{}/archive'.format(self.url, _get_path(entity_id))
+        if channel is not None:
+            url = '{}?channel={}'.format(url, channel)
+        return url
 
-    def file_url(self, entity_id, filename):
+    def file_url(self, entity_id, filename, channel=None):
         '''Generate a URL for a file in an archive without requesting it.
 
         @param entity_id The ID of the entity to look up.
         @param filename The name of the file in the archive.
+        @param channel optional channel information.
         '''
-        return '%s/%s/archive/%s' % (self.url, _get_path(entity_id), filename)
-
-    def files(self, entity_id, manifest=None, filename=None, read_file=False):
+        url = '{}/{}/archive/{}'.format(self.url, _get_path(entity_id),
+                                        filename)
+        if channel is not None:
+            url = '{}?channel={}'.format(url, channel)
```

Comment:
This logic is repeated multiple times. We should have it defined in a single place, and I like the urlencode approach you use above.